src/Draco.Compiler.Benchmarks/SyntaxBenchmarks.cs

```diff
@@ -0,0 +1,81 @@
+using System;
+using System.Collections.Generic;
+using BenchmarkDotNet.Attributes;
+using Draco.Compiler.Api.Syntax;
+using Draco.Compiler.Internal.Syntax;
+using SyntaxToken = Draco.Compiler.Internal.Syntax.SyntaxToken;
+
+namespace Draco.Compiler.Benchmarks;
+
+public class SyntaxBenchmarks : FolderBenchmarkBase
+{
+    private SyntaxToken[] tokens = null!;
+
+    public SyntaxBenchmarks()
+        : base("syntax")
+    {
+    }
+
+    [GlobalSetup]
+    public void Setup()
+    {
+        this.tokens = Lex(this.Input);
+    }
+
+    [Benchmark]
+    public void Lex()
+    {
+        var sourceReader = SourceReader.From(this.Input.Code);
+        var syntaxDiagnostics = new SyntaxDiagnosticTable();
+        var lexer = new Lexer(sourceReader, syntaxDiagnostics);
+
+        while (true)
+        {
+            var token = lexer.Lex();
+            if (token.Kind == TokenKind.EndOfInput) break;
+        }
+    }
+
+    [Benchmark]
+    public void Parse()
+    {
+        var tokenSource = TokenSource.From(this.tokens.AsMemory());
+        var syntaxDiagnostics = new SyntaxDiagnosticTable();
+        var parser = new Parser(tokenSource, syntaxDiagnostics);
+
+        _ = parser.ParseCompilationUnit();
```

Comment:
The JIT may do unwanted optimisation if we don't consume the output of a function.
You can just return the result as a whole, BDN accept it.
(Same comment for other benchmark methods)